src/payment/strategies/stripev3/stripev3-payment-strategy.ts

```diff
@@ -70,39 +72,58 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
-            .then(state => {
-                const paymentMethod = state.paymentMethods.getPaymentMethod(payment.methodId);
-
-                if (!paymentMethod || !paymentMethod.clientToken) {
-                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
-                }
-
-                if (!this._cardElement) {
-                    throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => {
+                if (paymentData) {
+                    return this._store.dispatch(this._paymentActionCreator.submitPayment({...payment, paymentData}));
                 }
 
-                return this._getStripeJs().handleCardPayment(
-                    paymentMethod.clientToken, this._cardElement, this._mapStripeCardPaymentOptions()
-                ).then(stripeResponse => {
-                    if (stripeResponse.error || !stripeResponse.paymentIntent.id) {
-                        throw new StandardError(stripeResponse.error && stripeResponse.error.message);
+                return this._paymentActionCreator.generatePaymentIntent(
+                    payment.methodId,
+                    shouldSaveInstrument,
+                    this._store,
+                    payment.gatewayId
+                ).then(paymentIntent => {
+                    if (!this._cardElement) {
+                        throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
                     }
 
-                    const paymentPayload = {
-                        methodId: payment.methodId,
-                        paymentData: { nonce: stripeResponse.paymentIntent.id },
-                    };
-
-                    return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
-                        .then(() =>
-                            this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload))
-                        );
+                    return this._getStripeJs().createPaymentMethod('card', this._cardElement,
+                        this._mapStripePaymentMethodOptions()
+                    ).then(stripePaymentMethod => {
+                        if (stripePaymentMethod.error || !stripePaymentMethod.paymentMethod.id) {
+                            throw new StandardError(stripePaymentMethod.error && stripePaymentMethod.error.message);
+                        }
+
+                        const stripeCardPaymentOptions = {
+                            ...this._mapStripeCardPaymentOptions(shouldSaveInstrument),
+                            payment_method: stripePaymentMethod.paymentMethod.id,
+                        };
+
+                        return this._getStripeJs().handleCardPayment(paymentIntent, stripeCardPaymentOptions)
+                            .then(stripeResponse => {
```

Comment:
This level of nesting can be removed by chaining with the outer promise instead. i.e.:

```ts
return this._getStripeJs().createPaymentMethod('card', this._cardElement, this._mapStripePaymentMethodOptions())
    .then(stripePaymentMethod => {
        // Do what you have to do with `stripePaymentMethod`
        return this._getStripeJs().handleCardPayment(paymentIntent, stripeCardPaymentOptions);
    })
    .then(stripeResponse => {
        // Do what you have to do with `stripePaymentMethod`
    });
```